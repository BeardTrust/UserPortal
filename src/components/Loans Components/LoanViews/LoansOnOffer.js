import React, { useCallback, useContext, useEffect, useState } from 'react';
import Pagination from '@material-ui/lab/Pagination';
import axios from "axios";
import { useHistory } from "react-router-dom";
import ActionContext from "../../../store/action-context";
import AuthContext from "../../../store/auth-context";
import { Table, Modal, Button } from "react-bootstrap"
import LoanRegistration from '../LoanSignUp/LoanRegistration';

function LoansOnOffer(props) {
    const history = useHistory();
    const actionContext = useContext(ActionContext);
    const authContext = useContext(AuthContext);
    const token = authContext.token;
    const userId = authContext.userId;
    //modal items
    const [applyLoan, setApplyLoan] = useState();
    const [show, setShow] = useState(false);
    const handleClose = () => setShow(false);
    //page items
    const [availableLoans, setavailableLoans] = useState();
    const [loansDisplayed, setLoansDisplayed] = useState(false);
    const [currentPage, setCurrentPage] = useState(1);
    const [pageSize, setPageSize] = useState(10);
    const [searchCriteria, setSearchCriteria] = useState("");
    const [sortBy, setSortBy] = useState("id,asc");
    const [searchCriteriaChanged, setSearchCriteriaChanged] = useState(false);
    const [sortByTypeName, setSortByTypeName] = useState({ active: false, name: 'typeName', direction: 'asc' });
    const [sortByDescription, setSortByDescription] = useState({ active: false, name: 'description', direction: 'asc' });
    const [sortByInterest, setSortByInterest] = useState({ active: false, name: 'baseInterestRate', direction: 'asc' });
    const url = 'http://localhost:9001/loantypes'
    const [numberOfPages, setNumberOfPages] = useState(10);
    const pageSizes = [5, 10, 15, 20, 25, 50, 100]

    const getList = useCallback(async () => {
        if (searchCriteria === "") {
            setSearchCriteria('');
        }
        console.log("search: ", searchCriteria);
        let params = null;
        if (searchCriteria !== '') {
            params = {
                page: currentPage === 0 ? 0 : currentPage - 1,
                size: pageSize,
                sortBy: sortBy,
                search: searchCriteria
            };
        } else {
            params = { page: currentPage === 0 ? 0 : currentPage - 1, size: pageSize, sortBy: sortBy };
        }

        console.log('params: ', params);
        const list = await axios.get(url, {
            params: params,
            headers: {
                'Authorization': token,
                'Content-Type': 'application/json'
            }
        });
        console.log("outbound url: ", url);
        console.log('inbound list: ', list)

        if (list.data.content !== availableLoans) {
            if (searchCriteriaChanged) {
                setCurrentPage(1);
                setSearchCriteriaChanged(false);
            }

            setLoansDisplayed(true);
            setavailableLoans(list.data.content);
            setNumberOfPages(list.data.totalPages);
        }
    }, [availableLoans, searchCriteriaChanged, token, pageSize, currentPage, searchCriteria, sortBy]);

    useEffect(() => {
        if (!loansDisplayed) {
            getList();
        }

    }, [getList, availableLoans, loansDisplayed, token, url]);

    function applyHandler() {
        console.log('attempting to apply')
        LoanRegistration(applyLoan, userId, token)
        handleClose();
    }

    function handlePageChange(event, value) {
        event.preventDefault();
        setLoansDisplayed(false);
        if (searchCriteriaChanged) {
            setCurrentPage(1);
        } else {
            setCurrentPage(value);
        }

    }

    function handleSearchCriteriaChange(event) {
        setSearchCriteriaChanged(true);
        setSearchCriteria(event.target.value);
    }

    function handlePageSizeChange(event) {
        setLoansDisplayed(false);
        setPageSize(event.target.value);
        setCurrentPage(1);
    }

    function addToSort(event) {
        let sort = '';
        let field = { };

        if (event.target.id === 'typeName') {
            if (sortByTypeName.active === true) {
                field = toggleDirection(sortByTypeName);
                sort = field.name + ',' + field.direction;
            } else {
                setSortByTypeName({ active: true, name: 'typeName', direction: 'asc' });
                sort = sortByTypeName.name + ',' + sortByTypeName.direction;
            }

            if (sortByDescription.active === true) {
                sort += ',' + sortByDescription.name + ',' + sortByDescription.direction;
            }

            if (sortByInterest.active === true) {
                sort += +',' + sortByInterest.name + ',' + sortByInterest.direction;
            }
        }

        if (event.target.id === 'description') {
            if (sortByDescription.active === true) {

                field = toggleDirection(sortByDescription);
                sort = field.name + ',' + field.direction;
            } else {
                setSortByDescription({ active: true, name: 'description', direction: 'asc' });
                sort = sortByDescription.name + ',' + sortByDescription.direction;
            }

            if (setSortByTypeName.active === true) {
                sort += +',' + setSortByTypeName.name + ',' + setSortByTypeName.direction;
            }

            if (sortByTypeName.active === true) {
                sort += ',' + sortByTypeName.name + ',' + sortByTypeName.direction;
            }
        }

        if (event.target.id === 'apr') {
            if (sortByInterest.active === true) {

                field = toggleDirection(sortByInterest);
                sort = field.name + ',' + field.direction;
            } else {
                setSortByInterest({ active: true, name: 'apr', direction: 'asc' });
                sort = sortByInterest.name + ',' + sortByInterest.direction;
            }

            if (sortByTypeName.active === true) {
                sort += ',' + sortByTypeName.name + ',' + sortByTypeName.direction;
            }

            if (sortByDescription.active === true) {
                sort += ',' + sortByDescription.name + ',' + sortByDescription.direction;
            }
        }

        setLoansDisplayed(false);
        setSortBy(sort);
    }

    function toggleDirection(field) {
        if (field.direction === 'asc') {
            field.direction = 'desc';
        } else {
            field.direction = 'asc';
        }

        return field;
    }

    return (
        <section className={'container'}>
            <h1 className={'text-center mt-5'}>The Loans of BeardTrust</h1>
            <div className={'mt-5'}>
                <div>
                    <div className={'input-group mb-3'}>
                        <div className={'me-5 col-xs-12 col-lg-2'}>
                            <span className={'align-middle'}>
                                {'Items per Page: '}
                            </span>
                            <select data-testid={'pageSizeSelector'} className={'text-center align-middle'} onChange={handlePageSizeChange}
                                value={pageSize}>
                                {pageSizes.map((size) => (
                                    <option key={size} value={size}>{size}</option>
                                ))}
                            </select>
                        </div>
                        <span className={'text-center col-sm-0 col-md-4 col-lg-6'} />
                        <input type={'text'} className={'form-control'} placeholder={'Search'} value={searchCriteria}
                            onChange={handleSearchCriteriaChange} />
                        <button className={'btn btn-outline-secondary'} type={'button'} onClick={getList}
                            id={'searchBar'}>Search
                        </button>
                    </div>
                </div>
                <Table striped bordered hover className={'me-3 table-responsive'} data-sortable={'true'}
                    data-toggle={'table'} id={'table'}>
                    <thead>
                        <tr>
                            <th className={'align-middle text-center'} data-sortable={'true'}
                                scope={'col'} onClick={addToSort}
                                id={'typeName'}>Loan
                                {sortByTypeName.active === true && (sortByTypeName.direction === 'asc' ? '  ↑' : '  ↓')}
                            </th>
                            <th data-sortable={'true'} scope={'col'} id={'description'} onClick={addToSort}>Description
                                {sortByDescription.active === true && (sortByDescription.direction === 'asc' ? '  ↑' : '  ↓')}
                            </th>
                            <th className={'align-middle text-center'} data-sortable={'true'} scope={'col'}
                                id={'apr'} onClick={addToSort}>Interest Rate
                                {sortByInterest.active === true && (sortByInterest.direction === 'asc' ? '  ↑' : '  ↓')}
                            </th>
                            <th className={'align-middle text-center'}>Sign Up</th>
                        </tr>
                    </thead>
                    <tbody>
                        {availableLoans && availableLoans.map(loan => (
                            <tr key={loan.id}>
                                <td className={'align-middle text-center'}>{loan.typeName}</td>
                                <td className={'align-middle'}>{loan.description}</td>
                                <td className={'align-middle text-center'}>{loan.apr + '%'}</td>
                                <td className={'align-middle text-center'}>
                                    <button className={'btn btn-primary btn mx-3'} onClick={function (event) { setApplyLoan(loan); setShow(true) }}
                                        id={loan.id}>Apply
                                    </button>
                                </td>
                            </tr>
                        ))}
                    </tbody>
                </Table>
                <Pagination className={'my-3'} count={numberOfPages} page={currentPage} siblingCount={1}
                    boundaryCount={1} onChange={handlePageChange} />
            </div>
            <script>$('#table').DataTable()</script>

            <Modal show={show} onHide={handleClose}>
                <Modal.Header closeButton>
                    <Modal.Title>Applying for Loan</Modal.Title>
                </Modal.Header>
                <Modal.Body>Loan Type: {applyLoan ? applyLoan.typeName : null}</Modal.Body>
                <Modal.Body>APR: {applyLoan ? applyLoan.apr : null}%</Modal.Body>
                <Modal.Body>Length: {applyLoan ? applyLoan.numMonths : null} months</Modal.Body>
                <Modal.Body>Description: {applyLoan ? applyLoan.description : null}</Modal.Body>
                <Modal.Footer>
                    <Button variant="secondary" onClick={handleClose}>
                        Cancel
                    </Button>
                    <Button variant="primary" onClick={() => applyHandler()}>
                        Apply for this Loan
                    </Button>
                </Modal.Footer>
            </Modal>
        </section>
    );
}

export default LoansOnOffer;