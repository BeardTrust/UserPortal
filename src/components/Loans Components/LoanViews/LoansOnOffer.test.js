
import TestRenderer from 'react-test-renderer'
import LoansOnOffer from './LoansOnOffer'
import {render, screen, act} from '@testing-library/react';
import axiosMock from 'axios';

jest.mock('axios')
describe('Loan View list', () => {


    const data = {
        data: {
            loanType: {
                id: 'cf8bd72a-4b4c-42f1-8dc3-06cc2dc2cb8a',
                typeName: 'auto',
                description: 'test',
                apr: 2.99,
                numMonths: 72,
                activeStatus: 'true'
            }
        }
    };

    it('Searches for loans', async () => {
        // const promise = Promise.resolve({status: 200});
        // axiosMock.get.mockResolvedValueOnce(promise);
        // const loanOff = render(<LoansOnOffer />);
        // const applyButton = screen.getByText('Search');
        // // await act(() => promise);
        // expect(axiosMock.get).toBeCalledTimes(1);
    });

    

    it('Expects the snapshot to match the display', async () => {
        const tr = TestRenderer.create(<LoansOnOffer/>);
        expect(tr.toJSON()).toMatchSnapshot();
    });
})