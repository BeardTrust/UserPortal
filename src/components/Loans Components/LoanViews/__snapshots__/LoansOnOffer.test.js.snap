// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loan View list Expects the snapshot to match the display 1`] = `
<section
  className="container"
>
  <h1
    className="text-center mt-5"
  >
    The Loans of BeardTrust
  </h1>
  <div
    className="mt-5"
  >
    <div>
      <div
        className="input-group mb-3"
      >
        <div
          className="me-5 col-xs-12 col-lg-2"
        >
          <span
            className="align-middle"
          >
            Items per Page: 
          </span>
          <select
            className="text-center align-middle"
            data-testid="pageSizeSelector"
            onChange={[Function]}
            value={10}
          >
            <option
              value={5}
            >
              5
            </option>
            <option
              value={10}
            >
              10
            </option>
            <option
              value={15}
            >
              15
            </option>
            <option
              value={20}
            >
              20
            </option>
            <option
              value={25}
            >
              25
            </option>
            <option
              value={50}
            >
              50
            </option>
            <option
              value={100}
            >
              100
            </option>
          </select>
        </div>
        <span
          className="text-center col-sm-0 col-md-4 col-lg-6"
        />
        <input
          className="form-control"
          onChange={[Function]}
          placeholder="Search"
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-secondary"
          id="searchBar"
          onClick={[Function]}
          type="button"
        >
          Search
        </button>
      </div>
    </div>
    <table
      className="me-3 table-responsive table table-striped table-bordered table-hover"
      data-sortable="true"
      data-toggle="table"
      id="table"
    >
      <thead>
        <tr>
          <th
            className="align-middle text-center"
            data-sortable="true"
            id="typeName"
            onClick={[Function]}
            scope="col"
          >
            Loan
          </th>
          <th
            data-sortable="true"
            id="description"
            onClick={[Function]}
            scope="col"
          >
            Description
          </th>
          <th
            className="align-middle text-center"
            data-sortable="true"
            id="apr"
            onClick={[Function]}
            scope="col"
          >
            Interest Rate
          </th>
          <th
            className="align-middle text-center"
          >
            Sign Up
          </th>
        </tr>
      </thead>
      <tbody />
    </table>
    <nav
      aria-label="pagination navigation"
      className="MuiPagination-root my-3"
    >
      <ul
        className="MuiPagination-ul"
      >
        <li>
          <button
            aria-label="Go to previous page"
            className="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page Mui-disabled Mui-disabled"
            disabled={true}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={-1}
            type="button"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiPaginationItem-icon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
              />
            </svg>
          </button>
        </li>
        <li>
          <button
            aria-current="true"
            aria-label="page 1"
            className="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page Mui-selected"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            1
            <span
              className="MuiTouchRipple-root"
            />
          </button>
        </li>
        <li>
          <button
            aria-label="Go to page 2"
            className="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            2
            <span
              className="MuiTouchRipple-root"
            />
          </button>
        </li>
        <li>
          <button
            aria-label="Go to page 3"
            className="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            3
            <span
              className="MuiTouchRipple-root"
            />
          </button>
        </li>
        <li>
          <button
            aria-label="Go to page 4"
            className="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            4
            <span
              className="MuiTouchRipple-root"
            />
          </button>
        </li>
        <li>
          <button
            aria-label="Go to page 5"
            className="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            5
            <span
              className="MuiTouchRipple-root"
            />
          </button>
        </li>
        <li>
          <div
            className="MuiPaginationItem-root MuiPaginationItem-ellipsis"
          >
            â€¦
          </div>
        </li>
        <li>
          <button
            aria-label="Go to page 10"
            className="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            10
            <span
              className="MuiTouchRipple-root"
            />
          </button>
        </li>
        <li>
          <button
            aria-label="Go to next page"
            className="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-page"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiPaginationItem-icon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
              />
            </svg>
            <span
              className="MuiTouchRipple-root"
            />
          </button>
        </li>
      </ul>
    </nav>
  </div>
  <script>
    $('#table').DataTable()
  </script>
</section>
`;
